test/fuse/linux/fuse_base.cc

```diff
@@ -0,0 +1,252 @@
+// limitations under the License.
+
+#include <iostream>
+
+#include <stdio.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/mount.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include <linux/fuse.h>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "fuse_base.h"
+
+namespace gvisor {
+namespace testing {
+
+void FuseTest::SetUp() {
+  MountFuse();
+  SetUpFuseServer();
+}
+
+void FuseTest::TearDown() { UnmountFuse(); }
+
+// Since CompareRequest is running in background thread, gTest assertions and
+// expectations won't directly reflect the test result. However, the FUSE
+// background server still connects to the same standard I/O as testing main
+// thread. So EXPECT_XX can still be used to show different results. To
+// ensure failed testing result is observable, return false and the result
+// will be sent to test main thread via pipe.
+bool FuseTest::CompareRequest(void* expected_mem, size_t expected_len,
+                              void* real_mem, size_t real_len) {
+  if (expected_len != real_len) return false;
+  return memcmp(expected_mem, real_mem, expected_len) == 0;
+}
+
+// SetExpected is called by the testing main thread to set expected request-
+// response pair of a single FUSE operation.
+int FuseTest::SetExpected(struct iovec* iov_in, int iov_in_cnt,
+                          struct iovec* iov_out, int iov_out_cnt) {
+  if (writev(set_expected_[1], iov_in, iov_in_cnt) <= 0) {
+    perror("writev iov_in failed");
+    return -1;
+  }
+
+  if (writev(set_expected_[1], iov_out, iov_out_cnt) <= 0) {
+    perror("writev iov_out failed");
+    return -1;
+  }
+
+  WaitCompleted();
+  return 0;
+}
+
+// WaitCompleted waits for the FUSE server to finish its job and check if it
+// completes without errors.
+void FuseTest::WaitCompleted() {
+  char success;
+  read(done_[0], &success, sizeof(success));
+  EXPECT_EQ(success, 1);
+}
+
+void FuseTest::MountFuse() {
+  char mount_opts[128];
+  dev_fd_ = open("/dev/fuse", O_RDWR);
+  ASSERT_GT(dev_fd_, 0);
+
+  sprintf(mount_opts, "fd=%d,%s", dev_fd_, MOUNT_OPTS);
+  ASSERT_EQ(
+      0, mount("fuse", MOUNT_POINT, "fuse", MS_NODEV | MS_NOSUID, mount_opts));
+}
+
+void FuseTest::UnmountFuse() {
+  EXPECT_EQ(umount(MOUNT_POINT), 0);
+  // TODO(gvisor.dev/issue/3330): ensure the process is terminated successfully
+}
+
+// ConsumeFuseInit consumes the first FUSE request and returns 0 if succeed.
+int FuseTest::ConsumeFuseInit() {
+  if (buf_ == nullptr) buf_ = malloc(FUSE_MIN_READ_BUFFER);
+
+  size_t in_len = sizeof(struct fuse_in_header) + sizeof(struct fuse_init_in);
+  if (read(dev_fd_, buf_, in_len) < 0) {
+    perror("read from /dev/fuse");
+    return -1;
+  }
+
+  struct iovec iov_out[2];
+  struct fuse_out_header out_header = {
+      .len = sizeof(struct fuse_out_header) + sizeof(struct fuse_init_out),
+      .error = 0,
+      .unique = 2,
+  };
+  // init out payload is ignored
```

Comment:
add period to the end of sentence, also do you mean "init out payload is going to be ignored by init"? since you are not ignoring it here.